include/inviwo/core/algorithm/permutations.h

```diff
@@ -0,0 +1,226 @@
+/*********************************************************************************
+ *
+ * Inviwo - Interactive Visualization Workshop
+ *
+ * Copyright (c) 2021 Inviwo Foundation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <inviwo/core/common/inviwocoredefine.h>
+
+#include <inviwo/core/util/transformiterator.h>
+#include <inviwo/core/util/exception.h>
+#include <tcb/span.hpp>
+
+#include <vector>
+#include <algorithm>
+
+namespace inviwo {
+
+namespace util {
+
+template <typename T>
+std::vector<T> makeVectorRange(ptrdiff_t start, ptrdiff_t end, ptrdiff_t step = 1) {
+    std::vector<T> res(static_cast<size_t>((end - start) / step));
+    std::generate(res.begin(), res.end(), [i = start, step]() mutable {
+        auto tmp = i;
+        i += step;
+        return static_cast<T>(tmp);
+    });
+    return res;
+}
+
+/**
+ * Generate all the successive \p r length permutations of the elements in \p values
+ * \see https://docs.python.org/3/library/itertools.html#itertools.permutations
+ *
+ * Example:
+ * \code
+ * std::array values = {0, 1, 2};
+ * util::Permutations perm(util::span{values}, 2);
+ * do {
+ *     for (auto v : perm) {
+ *         std::cout << v << " ";
+ *     }
+ *     std::cout << "\n";
+ * } while (perm.next());
+ * \endcode
+ */
+template <typename T>
+class Permutations {
```

Comment:
Do these new algorithms need to be in core?

Reply:
Yes, I needed them for the unit system